File: Programación Orientada a Objetos (PDOO)/Napakalaki(Java)/src/napakalaki/SpecificBadConsequence.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package napakalaki;

import java.util.ArrayList;

/**
 *
 * @author joseperoman
 */
public class SpecificBadConsequence extends BadConsequence {
    private ArrayList <TreasureKind> specificVisibleTreasures = new ArrayList();
    private ArrayList <TreasureKind> specificHiddenTreasures = new ArrayList();
    
    public SpecificBadConsequence(String txt, int lvl, ArrayList sv, ArrayList sh){
        super(txt,lvl);
        specificVisibleTreasures=sv;
        specificHiddenTreasures=sh;
    }
    
    public ArrayList <TreasureKind> getspecificVisibleTreasures(){
        return specificVisibleTreasures;
    }
    public ArrayList <TreasureKind> getspecificHiddenTreasures(){
        return specificHiddenTreasures;
    }
    
    @Override
    public boolean isEmpty(){
        boolean empty=false;
        if(specificVisibleTreasures.isEmpty() && specificHiddenTreasures.isEmpty()){
            empty=true;
        }
        return empty;
    }
    @Override
    public void substractVisibleTreasure(Treasure t){
        if(!specificVisibleTreasures.isEmpty()){
            specificVisibleTreasures.remove(t.getType());
        }
    }
    @Override
    public void substractHiddenTreasure(Treasure t){
        if(!specificHiddenTreasures.isEmpty()){
            specificHiddenTreasures.remove(t.getType());
        }
    }
    
    @Override
    public SpecificBadConsequence adjustToFitTreasureList(ArrayList<Treasure> v, ArrayList<Treasure> h){
        ArrayList<TreasureKind> igualesv= new ArrayList();
        ArrayList<TreasureKind> igualesh= new ArrayList();
        
        SpecificBadConsequence sbc = new SpecificBadConsequence("",0,igualesv,igualesh);
        
        if(!this.specificVisibleTreasures.isEmpty() || !this.specificHiddenTreasures.isEmpty()){
            for(int i=0; i<v.size(); i++){
                    if(this.specificVisibleTreasures.contains(v.get(i).getType())){
                        igualesv.add(v.get(i).getType());
                        i++;
                    }
                }
             for(int i=0; i<h.size(); i++){
                    if(this.specificHiddenTreasures.contains(h.get(i).getType())){
                        igualesh.add(h.get(i).getType());
                        i++;
                    }
                }
        }
       sbc= new SpecificBadConsequence(super.getText(),this.getLevels(),igualesv,igualesh);
        
        return sbc;
    }
    
    public String toString(){
        return /*" \n\tTexto = " + text + " \n\tLevels = " + Integer.toString(levels) + */"\n\tTesoros especificos visibles = " + specificVisibleTreasures + "\n\tTesoros especificos ocultos = " + specificHiddenTreasures + "\n";
    }
}

File: Programación Orientada a Objetos (PDOO)/Napakalaki(Ruby)/pkg/pruebaNapakalaki-0.0.1/lib/specific_bad_consequence.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

module NapakalakiGame
require_relative "bad_consequence.rb"

class SpecificBadConsequence < BadConsequence
  attr_accessor:specificVisibleTreasures,:specificHiddenTreasures 
  def initialize(txt,lvl,sv,sh)
    super(txt,lvl)
    @specificVisibleTreasures=sv
    @specificHiddenTreasures=sh
  end
  
  def isEmpty
    empty = false
    if @specificHiddenTreasures.size==0 && @specificVisibleTreasures.size==0 then
        empty=true
    end
    empty
  end
  
  def substractVisibleTreasures(t)      
    if (@specificVisibleTreasures.include?(t.type)) 
          @specificVisibleTreasures.delete(t.type)               
    end      
  end
  
  def substractHiddenTreasures(t)
    if (@specificHiddenTreasures.include?(t.type)) 
          @specificHiddenTreasures.delete(t.type)         
    end       
  end
=begin
   def adjustToFitTreasureLists(v, h)
    lH = Array.new
    lV = Array.new
    values = [TreasureKind::ARMOR, TreasureKind::ONEHAND, TreasureKind::BOTHHANDS, TreasureKind::HELMET, TreasureKind::SHOES]
        
    values.each { |tK|
      m1 = 0
      m2 = 0
                
      for i in 0..@specificVisibleTreasures.size-1
        if(@specificVisibleTreasures.at(i) == tK)
          m1 += 1
        end
      end
      for i in 0..v.size-1
        if(v.at(i).type == tK)
          m2 += 1
        end
      end
      for i in 0..[m1, m2].min-1
        lV.push(tK)
      end
                
      m1 = 0
      m2 = 0
                
      for i in 0..@specificHiddenTreasures.size-1
        if(@specificHiddenTreasures.at(i) == tK)
          m1 += 1
        end
      end
      for i in 0..h.size-1
        if(h.at(i).type() == tK)
          m2 += 1
        end
      end
      for i in 0..[m1, m2].min-1
        lH.push(tK)
      end
    }
        
    bc = SpecificBadConsequence.new(@text, 0, lV, lH)
     
    bc
  end
=end
  def adjustToFitTreasureList(v,h)
     igualesv= Array.new
     igualesh= Array.new
      bd = SpecificBadConsequence.new("", 0, 0, 0)

          if(!@specificVisibleTreasures.empty? || !@specificHiddenTreasures.empty?) then
           for i in 0..v.size()-1
               if(@specificVisibleTreasures.include?(v[i].type))
                   igualesv << v[i].type
                   i+= 1
               end
           end

           for i in 0..h.size()-1
               if(@specificHiddenTreasures.include?(h[i].type))
                   igualesh << (h[i].type)
                   i+=1
               end
           end

           bd = SpecificBadConsequence.new(@text, 0, igualesv, igualesh)
          end

      #bd.to_s
      bd    
  end

   def to_s
    "#{@text}\n Niveles perdidos: #{@levels}\n Visible especifico: #{@specificVisibleTreasures}\n Ocultos especificos: #{@specificHiddenTreasures}" 
  end
end

end